### PR TITLE
Associate data with the right TextEditor

### LINKED ISSUE #5: Uncaught TypeError: Cannot read property 'blockHeight' of undefined
[https://github.com/gandm/language-babel/issues/415](https://github.com/gandm/language-babel/issues/415)

Maybe a decorator.

### PR DESCRIPTION
**Pull request by Arcanemagus**
Make data associated with a specific `TextEditor`.

This primarily fixes two major bugs with the previous implementation:
* When Atom first started up, a request was sent off to calculate the sizes for all active `TextEditor`s, but the function used only worked on the currently active `TextEditor`. Modify this so that all data is associated with the proper `TextEditor`.

* When `import-cost` finished calculating the size of an import, the marker would be created on the correct line... but potentially the wrong `TextEditor` if the user had switched editors between when the request was initialized and when it was resolved.

Additionally:
* Make decorations click-through so clicking where they are changes the cursor line instead of doing nothing.
* Don't attempt to destroy the `Marker` in `onDidChange`: Although this is _correct_, it seems to be triggering an internal bug in Atom that I'm not sure whether is a misunderstanding on how to use `Markers` or a real bug in Atom. In either case, leaving the invalid `Marker` around till it can be destroyed when the `import-cost` results are updated shouldn't pose an issue.
* Handle the `done` even from `import-cost`. Although this shouldn't make a difference in almost every case, it's better to handle it than not.

This ended up being a larger change than I was thinking, feel free to use as much or as little of this as you would like @IanMitchell!

Fixes #5.

### CODE REVIEW

**Comment by IanMitchell** [comment]
Whoa, thank you so much for this! Looks awesome, and like a huge improvement. I'm still working through the changes and running it through the ringer, but seems pretty solid so far. Found one or two small bugs and a regression which I think results in a rare memory leak - will have more info soon 😄

**Comment by IanMitchell** [comment]
One issue I'm seeing - not sure why, but the text position is off when you switch to a tab for the first time. For long running processes, this can cause visual errors:

Visual

This is exacerbated by `import-cost` not finishing on itself ever, for whatever reason. Trying to look into that one now.

EDIT: To be clear, I'm looking into it and not asking you to fix it haha. Just wanted to let you know what I'm working on so it doesn't seem like I'm ignoring this. Little slow on fixing all of this since Atom plugin development is such a new environment for me

Edit 2: Looks like it's usually a blip, but a nasty memory leak made it take forever when I spotted it here. For whatever reason reloading the window isn't cleaning up import-cost threads - moving to fix that first

**Comment by IanMitchell** [comment]
Gonna merge in, then I can put up small fixes as I find them! Thanks again :D

**Comment by IanMitchell** [comment]
Seems to work in my 'old' repository - bet it was an issue with the native dependency being built / not built. Put one or two small tweaks in, gonna test for the day and then release!

**Comment by Arcanemagus** [comment]
> One issue I'm seeing - not sure why, but the text position is off when you switch to a tab for the first time.

That's what https://github.com/IanMitchell/import-cost-atom/pull/6/commits/7d4b90449a7c8dbac1b904638bc6393bf1ea862e was _supposed_ to fix (and worked for in my testing). An alternative to that is to check the return value of `getLineHeightInPixels` here:
https://github.com/IanMitchell/import-cost-atom/blob/cf0e95367bb3a9798e47ed862d38b6fb02f7497c/lib/import-cost-atom-view.js#L39

If it's `null` use a default value (that would then get updated on tab switching).

**Comment by Arcanemagus** [comment]
Also note that you are calling the same function twice here:

https://github.com/IanMitchell/import-cost-atom/blob/cf0e95367bb3a9798e47ed862d38b6fb02f7497c/lib/import-cost-atom-view.js#L38-L39

You should just move L39 above the `if` and use the value in the `if` condition 😉.

**Comment by Arcanemagus** [comment]
Ah ha! @IanMitchell it looks like the original problem _was_ a bug in Atom: https://github.com/atom/atom/pull/15865

~~Since it looks like that bug was in the stable channel I wouldn't revert https://github.com/IanMitchell/import-cost-atom/pull/6/commits/0846fea6e88f3e3783b15cdda0a4f9cb539dea2d though.~~

The fix was ported into v1.21.1, so if you want you can change this:
https://github.com/IanMitchell/import-cost-atom/blob/ff04c98ae527cdca600727097f2d8a7085e4f97e/lib/import-cost-atom-view.js#L22-L24

To:
```js
if (event.isValid) {
  this.updatePosition(editor);
} else {
  this.destroy();
}
```

**Comment by IanMitchell** [comment]
Recommend making the change haha? I'm a little unclear on what the difference would be now

**Comment by Arcanemagus** [comment]
So when the `Marker` becomes invalid that change would immediately remove the `ImportCostAtomView`. Currently the `Marker` becomes invalid (and thus hidden from the user), but is still in memory till `.destroy()` is called on the `ImportCostAtomView` instance. In other words it makes no difference to the user, and only as much difference to Atom as the memory usage of the `ImportCostAtomView` and associated objects (like the `Marker`).

tl;dr? It's more correct, but isn't really going to change anything 😛.

**Comment by IanMitchell** [comment]
Thanks, appreciate it! I'll probably make the change and just save it for the next release then, unless nothing else shows up in two weeks
